### ISSUE TITLE
Examples fail on macOS

### ISSUE THREAD

**Issue by seanisom** [wasi-gfx/wasi-gfx-runtime#2]
```
sean@Seans-MacBook-Pro example-runtime % cargo run -- --example triangle
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/example-runtime --example triangle`
thread 'tokio-runtime-worker' panicked at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-hal-0.18.1/src/metal/surface.rs:113:13:
get_metal_layer cannot be called in non-ui thread.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

May be an issue with the `wgpu` metal adapter - looking into it

**Comment by seanisom** [comment]
Full stacktrace - 
```
stack backtrace:
   0:        0x1064dde1c - std::backtrace_rs::backtrace::libunwind::trace::hadefa2fe489b32be
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:        0x1064dde1c - std::backtrace_rs::backtrace::trace_unsynchronized::h37e8bb3c3fd9c67c
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x1064dde1c - std::sys_common::backtrace::_print_fmt::h4994a90b70e9aaa7
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:68:5
   3:        0x1064dde1c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hb478ebbfb46e27ce
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:44:22
   4:        0x1064fa304 - core::fmt::rt::Argument::fmt::hfe2e626e52db21d5
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/fmt/rt.rs:142:9
   5:        0x1064fa304 - core::fmt::write::he4d5fa2daff1f531
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/fmt/mod.rs:1120:17
   6:        0x1064db840 - std::io::Write::write_fmt::hc5a47a68eba63d9f
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/io/mod.rs:1810:15
   7:        0x1064ddc50 - std::sys_common::backtrace::_print::hd47355bf70e1c2da
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:47:5
   8:        0x1064ddc50 - std::sys_common::backtrace::print::h79bd952cc5812e7a
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:34:9
   9:        0x1064df1dc - std::panicking::default_hook::{{closure}}::h82301f6222887737
  10:        0x1064def24 - std::panicking::default_hook::h1e49abbb3f1d7dbf
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:292:9
  11:        0x1064df624 - std::panicking::rust_panic_with_hook::h1e70c5d905e30e9d
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:779:13
  12:        0x1064df4ec - std::panicking::begin_panic_handler::{{closure}}::h399e32952efd26a4
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:649:13
  13:        0x1064de2a0 - std::sys_common::backtrace::__rust_end_short_backtrace::h2ab87f841a2323e7
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:171:18
  14:        0x1064df2a0 - rust_begin_unwind
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
  15:        0x10658f988 - core::panicking::panic_fmt::h33e40d2a93cab78f
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
  16:        0x104f06ff8 - wgpu_hal::metal::surface::<impl wgpu_hal::metal::Surface>::get_metal_layer::h6f981d7d1b5932ec
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-hal-0.18.1/src/metal/surface.rs:113:13
  17:        0x104f0690c - wgpu_hal::metal::surface::<impl wgpu_hal::metal::Surface>::from_view::hcd85a9b0dc33a331
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-hal-0.18.1/src/metal/surface.rs:87:34
  18:        0x104f09640 - <wgpu_hal::metal::Instance as wgpu_hal::Instance<wgpu_hal::metal::Api>>::create_surface::h9c298b7e423794a4
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-hal-0.18.1/src/metal/mod.rs:103:17
  19:        0x1047dc654 - wgpu_core::instance::<impl wgpu_core::global::Global<G>>::instance_create_surface::init::{{closure}}::hb87043602658924e
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-core-0.18.1/src/instance.rs:477:23
  20:        0x104aac9b4 - core::option::Option<T>::and_then::h068b8e5860a5ef25
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/option.rs:1406:24
  21:        0x1047dc5e4 - wgpu_core::instance::<impl wgpu_core::global::Global<G>>::instance_create_surface::init::h6e3692b6dd238cdd
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-core-0.18.1/src/instance.rs:476:13
  22:        0x104ad8a70 - wgpu_core::instance::<impl wgpu_core::global::Global<G>>::instance_create_surface::h9b4e956802fda143
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-core-0.18.1/src/instance.rs:495:20
  23:        0x10496a23c - example_runtime::webgpu::<impl example_runtime::component::webgpu::webgpu::HostGpuDevice for example_runtime::HostState>::connect_graphics_context::h308e3344779a27b2
                               at /Users/sean/code/wasi-webgpu/example-runtime/src/webgpu.rs:27:23
  24:        0x104a54d4c - example_runtime::component::webgpu::webgpu::add_to_linker::{{closure}}::hed895243044cff46
                               at /Users/sean/code/wasi-webgpu/example-runtime/src/main.rs:52:1
  25:        0x104b8b3b0 - wasmtime::component::func::host::HostFunc::entrypoint::{{closure}}::{{closure}}::h75669e0bf14dd700
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-17.0.0/src/component/func/host.rs:68:35
  26:        0x104bc053c - wasmtime::component::func::host::call_host::h36bf29759499cee9
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-17.0.0/src/component/func/host.rs:227:15
  27:        0x104b86e68 - wasmtime::component::func::host::HostFunc::entrypoint::{{closure}}::hd5895fe07976f6ae
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-17.0.0/src/component/func/host.rs:60:17
  28:        0x104c6fe00 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h4a7ce558cd869df6
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panic/unwind_safe.rs:272:9
  29:        0x1047fe814 - std::panicking::try::do_call::hbee55136f021170d
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:552:40
  30:        0x10480d41c - ___rust_try
  31:        0x1047ee96c - std::panicking::try::h98895830f8961c09
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:516:19
  32:        0x1048df0f8 - std::panic::catch_unwind::h23e2fa1e1bc39c38
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panic.rs:142:14
  33:        0x104b591c8 - wasmtime::component::func::host::handle_result::hf2e6052778d73215
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-17.0.0/src/component/func/host.rs:305:11
  34:        0x104b765d8 - wasmtime::component::func::host::HostFunc::entrypoint::h573e9d5632e8a5ac
                               at /Users/sean/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-17.0.0/src/component/func/host.rs:59:13
  35:        0x10dfe10ac - <unknown>
  ```

Not entirely clear in the stacktrace due to the wasmtime invocation, but I think what is happening is that the callback to the example host for the `webgpu` component calls is happening from the thread that is spawned to run the wasm payload: https://github.com/MendyBerger/wasi-webgpu/blob/main/example-runtime/src/main.rs#L301

Apparently this isn't going to work for `metal`, as it appears that initialization of the `HostGpuDevice` must happen on the main thread: https://github.com/gfx-rs/wgpu/blob/2382c8e74f44a3b0a26abeede18c70ec817e3394/wgpu-hal/src/metal/surface.rs#L102

**Comment by MendyBerger** [comment]
We can probably send over this call to the main thread.

**Comment by MendyBerger** [comment]
Fixed in #38
